Indexing/Index.hpp

```diff
@@ -19,102 +19,124 @@
 #include "Debug/Output.hpp"
 
 #include "Lib/Event.hpp"
+#include "Kernel/Clause.hpp"
+#include "Kernel/Term.hpp"
 #include "Lib/Exception.hpp"
 #include "Lib/VirtualIterator.hpp"
 #include "Saturation/ClauseContainer.hpp"
 #include "ResultSubstitution.hpp"
 #include "Kernel/UnificationWithAbstraction.hpp"
-
 #include "Lib/Allocator.hpp"
 
+/**
+ * Indices are parametrized by a LeafData, i.e. the bit of data you want to store in the index.
+ * Each leaf data must have a key to store at the leave. The Key can currently be either a Literal* or a TypedTermList.
+ * A LeafData must have a function  `<Key> key() const;` returns the key, and must have comparison operators (<=,<,>=,>,!=,==) implemented.
+ * See e.g. TermLiteralClause below for examples.
+ */
 namespace Indexing
 {
 using namespace Kernel;
 using namespace Lib;
 using namespace Saturation;
 
-
-/**
- * Class of objects which contain results of single literal queries.
- */
-template<class Unifier>
-struct LQueryRes
+struct LiteralClause 
 {
-  LQueryRes() {}
-  LQueryRes(Literal* l, Clause* c, Unifier unifier)
-  : literal(l), clause(c), unifier(std::move(unifier)) {}
+  Literal* const& key() const
+  { return literal; }
 
+private:
+  auto asTuple() const
+  { return std::make_tuple(clause->number(), literal->getId()); }
+public:
 
-  Literal* literal;
-  Clause* clause;
-  Unifier unifier;
+  IMPL_COMPARISONS_FROM_TUPLE(LiteralClause)
+
+  Literal* literal = nullptr;
+  Clause* clause = nullptr;
+
+  friend std::ostream& operator<<(std::ostream& out, LiteralClause const& self)
+  { return out << "{ " << outputPtr(self.clause) << ", " << outputPtr(self.literal) << " }"; }
+};
+
+template<class Value>
+struct TermWithValue {
+  TypedTermList term;
+  Value value;
+
+  TermWithValue() {}
+
+  TermWithValue(TypedTermList term, Value v)
+    : term(term)
+    , value(std::move(v))
+  {}
+
+  TypedTermList const& key() const { return term; }
+
+  auto asTuple() const
+  { return std::tie(term, value); }
```

Comment:
Yes exactly this is the reason why we're using the tie here. We can't do this in LiteralClause because we can't return a reference to the result of clause->number() and literal->getId() (as they are only live during the call of asTuple), thus we need to use make_tuple in this case. In any case we want as tuple just to return some tuple of which evey element can be ordered using std::less, hashed, ..., and what exact the type tuple elements are doesn't matter.